django/contrib/auth/management/__init__.py

```diff
@@ -111,47 +113,73 @@ def create_permissions(
 
 
 class RenamePermission(migrations.RunPython):
-    def __init__(self, app_label, old_model, new_model):
+    def __init__(self, app_label, old_model, new_model, backward, using):
         self.app_label = app_label
         self.old_model = old_model
         self.new_model = new_model
-        super(RenamePermission, self).__init__(
-            self.rename_forward, self.rename_backward
-        )
+        self.backward = backward
+        self.using = using
+        super().__init__(self._rename, self._rename)
+
+    def _rename(self, apps, schema_editor):
+        if self.backward:
+            from_model, to_model = self.new_model, self.old_model
+            model_class = apps.get_model(self.app_label, from_model)
+        else:
+            from_model, to_model = self.old_model, self.new_model
+            model_class = apps.get_model(self.app_label, to_model)
 
-    def _rename(self, apps, schema_editor, old_model, new_model):
-        ContentType = apps.get_model("contenttypes", "ContentType")
         # Use the live Permission model instead of the frozen one, since frozen
         # models do not retain foreign key constraints.
         from django.contrib.auth.models import Permission
 
-        db = schema_editor.connection.alias
-        ctypes = ContentType.objects.using(db).filter(
-            app_label=self.app_label, model__icontains=old_model.lower()
+        perms = Permission.objects.using(self.using).filter(
+            content_type__app_label=self.app_label,
+            codename__in=[
+                action + "_" + from_model.lower()
+                for action in model_class._meta.default_permissions
+            ],
         )
-        for permission in Permission.objects.using(db).filter(
-            content_type_id__in=ctypes.values("id")
-        ):
-            prefix = permission.codename.split("_")[0]
-            default_verbose_name = camel_case_to_spaces(new_model)
-
-            new_codename = f"{prefix}_{new_model.lower()}"
-            new_name = f"Can {prefix} {default_verbose_name}"
-
-            if permission.codename != new_codename or permission.name != new_name:
-                permission.codename = new_codename
-                permission.name = new_name
+        for permission in perms:
+            action = permission.codename.split("_")[0]
+            new_codename = action + "_" + to_model.lower()
+            if self.backward:
+                # We don't have the target model class, so approximate.
+                # TODO: might be able to get this if we adjust the order of
+                # injected operations when migrating backwards.
+                verbose_name_raw = camel_case_to_spaces(from_model)
+            else:
+                verbose_name_raw = model_class._meta.verbose_name_raw
+            new_name = f"Can {action} {verbose_name_raw}"
```

Comment:
In the forward direction, this now uses the verbose_name of the model.

Next week, I can look into the TODO I left about getting this in the backward direction.

Once that's resolved, it should be more clear to how to tighten up the `perms` query above to make sure it's not over-fetching (and trying to reconcile) identically named permissions that are tied to different content types.